bot/cogs/regular_expressions.py

```diff
@@ -0,0 +1,126 @@
+
+    Which will look as:
+    |    hello123456world
+    | 0:      ^^^^^^
+    | 1:          ^^
+    """
+    if match is None:
+        return ["No match"]
+    else:
+        group_carets = []
+        captured_positions = match.regs
+
+        for group_index, (start, end) in enumerate(captured_positions):
+            group_is_missing = (start, end) == (-1, -1)
+            if not group_is_missing:
+                group_carets.append(
+                    f"{group_index:>2}: " + " "*start + "^"*(end - start)
+                )
+
+        return ["    " + match.string, *group_carets]
+
+
+def match_and_format(pattern: Union[regex.Regex], test: str) -> str:
+    """Attempt to match a regex with a test string and return a formatted result."""
+    match_lines = "\n".join(format_match(pattern.search(test)))
+    return f"```\n{match_lines}\n```"
+
+
+class ConvertRegex(Converter):
+    """
+    Argument converter.
+
+    Attempts to interpret the given string as a regex
+    and returns a compiled pattern as a result.
+    """
+
+    def __init__(self, supports_extended_features: bool = False):
+        self.supports_extended_features = supports_extended_features
+
+    async def convert(self, ctx: Context, supposed_regex: str) -> regex.Regex:
+        """
+        Attempt to parse the input string as a regular expression.
+
+        If `self.supports_extended_features` is False, then the regex is forced
+        to be compatible with `re`.
+        """
+        try:
+            if not self.supports_extended_features:
+                re.compile(supposed_regex)
+            return regex.compile(supposed_regex)
+        except (regex.error, re.error) as e:
+            error_lines = "\n".join(format_error(e))
+            raise BadArgument(
+                "\nSyntax error in a regular expression: "
+                f"```\n{error_lines}\n```"
+            )
+
+
+ReRegex = ConvertRegex(supports_extended_features=False)
+RegexRegex = ConvertRegex(supports_extended_features=True)
+
+
+class RegularExpressions(Cog):
+    """Commands related to regular expressions."""
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+
+    @group(name='regexp', aliases=('regex', 're'), invoke_without_command=True)
+    async def regexp_group(self, ctx: Context) -> None:
+        """Commands for exploring the misterious world of regular expressions."""
+        await ctx.invoke(self.bot.get_command("help"), "regexp")
+
+    @regexp_group.command(name='search', aliases=('find', 's', '🔍'))
+    async def match_command(self, ctx: Context, pattern: ReRegex, test: str) -> None:
+        """Look for the first match of a pattern in a string."""
+        await ctx.send(match_and_format(pattern, test))
+
+    @regexp_group.command(name='search+', aliases=('find+', 's+', '🔍+'))
+    async def match_plus_command(self, ctx: Context, pattern: RegexRegex, test: str) -> None:
+        """
+        Like `!re search`, but with an extended regex format.
```

Comment:
It's a bad user experience to make them have to invoke the help command again to look at the docstring for `!re search`. Just copy the docstring over to this one.